### PR TITLE
Add ‘transaction_version’ to the signed transaction

### LINKED ISSUE #5951: `transaction_version` is not part of the transaction, rendering hardware wallets insecure
Hardware wallets guarantee that their display is a faithful representation of the transaction they are about to sign.  To ensure this, they need to be able to parse the transaction before they sign it.

New versions of hardware wallet firmware require manual approval and signing before they can be used.  This process generally takes weeks.  Therefore, it is not possible to update the firmware whenever the runtime is upgraded.  Instead, each transaction must have an associated version, so that a signed transaction blob will be rejected by an incompatible runtime, rather than being misinterpreted.

Currently, this is handled by the `spec_version` of the runtime.  However, this is changed frequently ― far more frequently than hardware wallets can be updated.  To solve this, there needs to be a subset of the runtime with a long term stability guarantee.  That is what `transaction_version` is for (https://github.com/paritytech/substrate/pull/5582).

However, `transaction_version` is currently not part of the transaction itself.  This renders hardware wallets insecure: an attacker can submit an unsigned transaction blob with a transaction version that it was not intended for, causing the transaction to be misinterpreted.  This breaks the guarantees provided by the hardware wallet.

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
This allows hardware wallets to know which transactions they can safely
sign.  To reduce transaction size, I reduced it to a ‘u8’ from a ‘u32’.

Fixes #5951.

### CODE REVIEW

**Comment by jacogr** [comment]
Without a TRANSACTION_VERSION bump this breaks a lot of code out in the wild, anything that parses old blocks now will fail since the format is unexpected. This is especially apparent cases where the signed extensions are not yet available in the metadata.

Would not rush this to putting on-chain at all since there are a lot of updates that should be made to wallets, extensions etc. to cater for the this, specifically -

- the new signed extensions should be added to APIs
- the APIs should adjust their payload generation for the new fields
- wallets, extensions should update to these
- work-arounds for detecting failures of non-matching formats in older blocks

File: primitives/version/src/lib.rs [hunk 1]
```diff
@@ -103,7 +103,7 @@ pub struct RuntimeVersion {
 	/// index.
 	///
 	/// It need *not* change when a new module is added or when a dispatchable is added.
-	pub transaction_version: u32,
+	pub transaction_version: u8,
```
Comment (bkchr) [inline primitives/version/src/lib.rs:106]: Nope, this will break all chains. Please revert.

Comment (DemiMarie-parity) [reply primitives/version/src/lib.rs:106]: Will do.  Is the rest of the code okay?

**Comment by DemiMarie-parity** [comment]
@bkchr Will `transaction_version` still be checked even if it is not included in the tx?

**Comment by jacogr** [comment]
The additional data form part of the payload that gets signed. If the signature gets verified, these are supplied by the runtime. When there is a mismatch, the signature verification fails. (This is exactly how spec_version operates)

**Comment by kirushik** [comment]
@kianenigma as far as I understand, we only want to check for semantical `transaction_version` most of the time — runtime might be upgraded while keeping the `transaction_version` intact, and our signatures (coming from an offline device knowing nothing about the recent runtime upgrades) should still be valid.

**Comment by bkchr** [comment]
> @bkchr Will `transaction_version` still be checked even if it is not included in the tx?

You have implemented it, please read the docs if you are unsure what you are doing here.

**Comment by bkchr** [comment]
> @kianenigma as far as I understand, we only want to check for semantical `transaction_version` most of the time — runtime might be upgraded while keeping the `transaction_version` intact, and our signatures (coming from an offline device knowing nothing about the recent runtime upgrades) should still be valid.
> 
> _Edit:_ I'm no longer sure if only checking `transaction_version` and not `spec_version` is a good idea. It's up to the particular chain implementation to decide on this, though.

The offline device still needs to append the `spec_version` when signing the transaction. It can be passed in the payload that will be signed, but it needs to be signed.

File: frame/system/src/lib.rs [hunk 2]
```diff
@@ -53,7 +53,9 @@
 //!   - [`CheckEra`]: Checks the era of the transaction. Contains a single payload of type `Era`.
 //!   - [`CheckGenesis`]: Checks the provided genesis hash of the transaction. Must be a part of the
 //!     signed payload of the transaction.
-//!   - [`CheckVersion`]: Checks that the runtime version is the same as the one encoded in the
+//!   - [`CheckSpecVersion`]: Checks that the runtime version is the same as the one encoded in the
+//!     transaction.
+//!   - [`CheckTxVersion`]: Checks that the transaction version is the same as the one encoded in the
 //!     transaction.
```
Comment (bkchr) [inline frame/system/src/lib.rs:59]:
```suggestion
//!   - [`CheckSpecVersion`]: Checks that the runtime version is the same as the one used to sign the
//!     transaction.
//!   - [`CheckTxVersion`]: Checks that the transaction version is the same as the one used the
//!     transaction.
```

File: frame/system/src/lib.rs [hunk 3]
```diff
@@ -1656,14 +1658,49 @@ impl<T: Trait + Send + Sync> SignedExtension for CheckGenesis<T> {
...
+	/// Create new `SignedExtension` to check transaction version.
+	pub fn new() -> Self {
+		Self(sp_std::marker::PhantomData)
+	}
+}
+
+impl<T: Trait + Send + Sync> SignedExtension for CheckTxVersion<T> {
+	type AccountId = T::AccountId;
+	type Call = <T as Trait>::Call;
+	type AdditionalSigned = u8;
```
Comment (bkchr) [inline frame/system/src/lib.rs:1687]:
```suggestion
	type AdditionalSigned = u32;
```

**Comment by DemiMarie-parity** [comment]
@bkchr can you re-review?

File: bin/utils/subkey/src/main.rs [hunk 4]
```diff
@@ -716,6 +717,7 @@ fn create_extrinsic<C: Crypto>(
 		extra(index, 0),
 		(
 			VERSION.spec_version as u32,
```
Comment (andresilva) [inline bin/utils/subkey/src/main.rs:719]: This `as u32` is probably not needed.

**Comment by DemiMarie-parity** [comment]
@bkchr can you re-review this?

**Comment by xlc** [comment]
I don't get this. spec_version is already included in tx signing payload which means any runtime upgrade that bumps of spec_version will invalidate all transactions signed using old spec_version.

Update transaction_version by definition will need to update spec_version.

It will make sense to only check transaction_version so spec_version bump will not necessarily invalidate hardware wallets signing ability without upgrade.

**Comment by kirushik** [comment]
@xlc the expectation is that chains would want _both_ behaviours in most of the cases — this is why it makes sense to provide them both, while letting a particular chain to pick any combination they like.

`transaction_version` is for offline wallets, bumping it it designates breaking semantic changes between runtime version. If it changes, wallet can no longer be sure that it parsed and presented the payload in a way that will be accepted by the chain. At the same time, wallet is expected to sign with any `spec_version` it is being asked about, and shouldn't care.

`spec_version` is an always-incremented version counter of the current runtime. Most chains would still want to invalidate the transaction pool after runtime upgrade, even if no _semantic_ changes happened between versions.

Take the recent [Kusama redenomination proposal](https://kusama.polkassembly.io/referendum/52) — it will mess with balances and values, and because of that we want all the previous transactions to be rendered invalid; at the same time, there's nothing in there which breaks the transaction parsing semantics in that upgrade, so invalidating say Ledger firmware during that upgrade would not win us anything.

Did I misinterpreted your question?

**Comment by jacogr** [comment]
It is not a Kusama redomination proposal - Kusama stays as-is. It is a sentiment proposal for Polkadot.

**Comment by xlc** [comment]
It still does not make sense for a chain to include both check spec and check tx, which is this PR and polkadot one does.

**Comment by bkchr** [comment]
What? Where is this a default? You can just continue to use the `CheckSpecVersion` extension.

**Comment by jacogr** [comment]
Will explain where this comes from -

- signers (Ledger, Parity Signer) only receives the actual payload, nothing more
- you don't want to display the hex value to the user, but rather the decoded value
- specVersion doesn't indicate that nothing in the tx has changed, so txVersion was introduced
- here the txVersion is put in the payload, so signers can make a call whether to use the existing metadata (even if there has been an upgrade) for known methods (new ones won't be decodable)

The alternative is to push a new app version for approval on each upgrade. It will be done anyway, but it means that users can continue to sign (assuming no txVersion updates) before approvals.

This is not driven by chain requirement, this is driven by security for signers.

It would be great if metadata can be pushed to signers without store approvals, but we are not there everywhere atm. (And obviously, it needs to be from a trusted source).

**Comment by xlc** [comment]
ah thanks for the explanation @jacogr, it makes sense for me now

@bkchr by default I mean node-template, it is the default template for new projects.

**Comment by bkchr** [comment]
@xlc yeah that is right, but that should be used as a starting point anyway. I would expect that after getting familiar with Substrate, you start thinking of what your chain needs and what not. ;)

**Pull request merged**